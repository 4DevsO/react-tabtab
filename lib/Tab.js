'use strict';

var React = require('react');
var classNames = require('classnames');
var Tappable = require('react-tappable');

var Tab = React.createClass({
  displayName: 'Tab',

  handleTabDeleteButton: function handleTabDeleteButton() {
    this.props.handleTabDeleteButton(this.props.tabKey);
  },

  clickTab: function clickTab() {
    this.props.handleTabClick(this.props.tabKey);
  },

  render: function render() {
    var tabClass;
    var tabTmpl;

    if (this.props.status === 'active') {
      tabClass = classNames(this.props.style + 'tab', 'active');
    } else {
      tabClass = classNames(this.props.style + 'tab');
    }
    if (this.props.tabKey === 'ADD') {
      tabClass = classNames(this.props.style + 'tab', 'add');
    }

    // only show the delete button when it's active
    var closeButtonStyle;
    if (this.props.tabDeleteButton && this.props.status === "active") {
      closeButtonStyle = { display: 'inline-block' };
    } else {
      closeButtonStyle = { display: 'none' };
    }

    return React.createElement(
      Tappable,
      { onClick: this.clickTab },
      tabTmpl,
      React.createElement(
        'li',
        { className: tabClass },
        this.props.title,
        React.createElement(
          'div',
          { className: 'tabtab__folder__circCont', style: closeButtonStyle },
          React.createElement('div', { className: 'tabtab__folder__circle', onClick: this.handleTabDeleteButton })
        )
      )
    );
  }
});

module.exports = Tab;