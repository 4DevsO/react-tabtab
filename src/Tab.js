import React from 'react';
import styled from 'styled-components';

const TabStyle = styled.div`
  display: inline-block;
  color: ${props => props.active ? '#007bff' : 'black'};
  border-bottom: ${props => props.active ? '2px solid #007bff' : '1px solid #eee'};
  padding: 5px 10px;
  transition: color .3s cubic-bezier(.645, .045, .355, 1);
  &:hover {
    cursor: pointer;
    color: #007bff;
  }
`;

export default class Tab extends React.Component {
  constructor(props) {
    super(props);
    this.clickTab = this.clickTab.bind(this);
  }

  clickTab() {
    const {handleActiveIndex, index} = this.props;
    console.log(index)
    handleActiveIndex(index);
  }

  render() {
    const {customStyle, active} = this.props;
    // console.log(this.props)
    const TabComponent = customStyle || TabStyle;
    return (
      <TabComponent onClick={this.clickTab} active={active}>
        {this.props.children}
      </TabComponent>
    )
  }
}

export {
  TabStyle
};