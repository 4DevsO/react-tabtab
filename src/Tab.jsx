import React from 'react';
import ReactDOM from 'react-dom';
import classNames from 'classnames';
import Tappable from 'react-tappable';
import ItemTypes from './ItemTypes';
import { DragSource, DropTarget } from 'react-dnd';

const tabSource = {
  beginDrag: function(props) {
    props.beginDrag();
    return {
      tabKey: props.tabKey
    }
  }
}

// reference from https://github.com/gaearon/react-dnd/tree/master/examples/04%20Sortable/Simple
const tabTarget = {
  hover(props, monitor, component) {
    const dragIndex = monitor.getItem().tabKey;
    const hoverIndex = props.tabKey;

    // Don't replace items with themselves
    if (dragIndex === hoverIndex) {
      return;
    }

    // Determine rectangle on screen
    const hoverBoundingRect = ReactDOM.findDOMNode(component).getBoundingClientRect();

    // Get horizontal middle
    const hoverMiddleX = (hoverBoundingRect.right - hoverBoundingRect.left) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    const hoverClientX = hoverBoundingRect.right - clientOffset.x;

    // Dragging left
    if (dragIndex < hoverIndex && hoverClientX < hoverMiddleX) {
      return;
    }

    // Dragging right
    if (dragIndex > hoverIndex && hoverClientX > hoverMiddleX) {
      return;
    }

    props.moveTab(dragIndex, hoverIndex);
    // change the dragging tab index
    monitor.getItem().tabKey = hoverIndex;
  }
};

@DropTarget(ItemTypes.TAB, tabTarget, connect => ({
  connectDropTarget: connect.dropTarget()
}))
@DragSource(ItemTypes.TAB, tabSource, (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  isDragging: monitor.isDragging()
}))

export default class Tab extends React.Component {
  constructor(props) {
    super(props);
    this.handleTabDeleteButton = this.handleTabDeleteButton.bind(this);
    this.clickTab = this.clickTab.bind(this);
  }

  handleTabDeleteButton(e) {
    e.stopPropagation(); //prevent trigger clickTab function
    this.props.handleTabDeleteButton(this.props.tabKey);
  }

  clickTab() {
    this.props.handleTabClick(this.props.tabKey);
  }

  render() {
    let tabClass,
        closeButtonStyle;
    const {connectDragSource, connectDropTarget } = this.props;
    if (this.props.status === 'active') {
      tabClass = classNames(this.props.style + 'tab', 'active');
    } else {
      tabClass = classNames(this.props.style + 'tab');
    }

    // only show the delete button when it's active
    if (this.props.tabDeleteButton && this.props.status === "active") {
      closeButtonStyle = {display: 'inline-block'};
    } else {
      closeButtonStyle = {display: 'none'};
    }


    return connectDragSource(connectDropTarget(
      <span>
        <li className={tabClass} onClick={this.clickTab}>
          {this.props.title}
          <div className="tabtab__folder__circCont" style={closeButtonStyle}>
            <div className="tabtab__folder__circle" onClick={this.handleTabDeleteButton}></div>
          </div>
        </li>
      </span>
    ));
  }
}