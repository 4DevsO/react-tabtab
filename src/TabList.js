import React from 'react';
import styled from 'styled-components';
import MdChevronLeft from 'react-icons/lib/md/chevron-left';
import MdChevronRight from 'react-icons/lib/md/chevron-right';

const ListWrapper = styled.div`
  position: relative;
  white-space: nowrap;
  overflow: hidden;
  width: auto;
  border-bottom: 1px solid #eee;
  padding: ${props => props.hasExtraButton ? '0 32px' : '0'};
`;

const ListStyle = styled.div`
  overflow: hidden;
`;

const ListScroll = styled.div`
  padding-left: 0;
  position: relative;
  margin: 0;
  list-style: none;
  display: inline-block;
  transition: transform .3s cubic-bezier(.42, 0, .58, 1);
`;

const ScrollButton = styled.div`
  display: inline-block;
  filter: none;
  position: absolute;
  ${props => props.left ?
    'left: 0'
  : 'right: 0'
  };
  &:hover {
    cursor: pointer;
  }
`;

export default class TabList extends React.Component {
  constructor(props) {
    super(props);
    this.handleScroll = this.handleScroll.bind(this);
    this.scrollPosition = 0;
    this.tabRefs = [];
  }

  getTabNode(tab) {
    const tabType = tab.constructor.name;
    if (tabType === 'Tab') {
      return tab.tab.__INTERNAL_NODE;
    } else if (tabType === 'DragTab') {
      return tab.__INTERNAL_NODE.node;
    }
  }

  unifyScrollMax(width) {
    return parseFloat((width / 3) * 2);
  }

  handleScroll(direction) {
    let leftMove = 0;
    const containerOffset = this.listContainer.getBoundingClientRect();
    const containerWidth = this.listContainer.offsetWidth;
    const tabFirstOffset = this.getTabNode(this.tabRefs[0]).getBoundingClientRect();
    const tabLastOffset = this.getTabNode(this.tabRefs[this.tabRefs.length - 1]).getBoundingClientRect();
    if (direction === 'right') {
      leftMove = tabLastOffset.right - containerOffset.right;
      if (leftMove > containerWidth) {
        leftMove = this.unifyScrollMax(containerWidth);
      }
      this.scrollPosition += leftMove;
    } else if (direction === 'left') {
      leftMove = tabFirstOffset.left - containerOffset.left;
      if (-leftMove > containerWidth) {
        leftMove = - this.unifyScrollMax(containerWidth);
      }
      this.scrollPosition += leftMove;
    }
    if (this.scrollPosition < 0) {
      this.scrollPosition = 0;
    }

    this.listScroll.style.transform = `translate3d(-${this.scrollPosition}px, 0, 0)`;
  }

  render() {
    const {
      customStyle,
      children,
      activeIndex,
      handleTabChange,
      handleEdit,
      ExtraButton
    } = this.props;
    const props = {
      handleTabChange,
      handleEdit
    };
    const ListInner = customStyle || ListStyle;
    return (
      <div>
        {ExtraButton ? ExtraButton : null}
        <ListWrapper hasExtraButton={!!ExtraButton}>
          <ScrollButton left onClick={() => {this.handleScroll('left')}}>
            <MdChevronLeft/>
          </ScrollButton>
          <ScrollButton onClick={() => {this.handleScroll('right')}}>
            <MdChevronRight/>
          </ScrollButton>
          <ListInner innerRef={node => this.listContainer = node}>
            <ListScroll innerRef={node => this.listScroll = node}>
              {React.Children.map(children, (child, index) => (
                React.cloneElement(child, {
                  key: index,
                  active: index === activeIndex,
                  index,
                  tabIndex: index,
                  ref: node => {
                    this.tabRefs.push(node)
                  },
                  ...props
                })
              ))}
            </ListScroll>
          </ListInner>
        </ListWrapper>
      </div>
    )
  }
}

TabList.displayName = 'TabList';

export {
  ListStyle
};