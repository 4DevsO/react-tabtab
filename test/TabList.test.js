import React from 'react';
import TabList, { ListStyle } from '../src/TabList';
import Tab from '../src/Tab';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import 'jest-styled-components'
import MdFormatListBulleted from 'react-icons/lib/md/format-list-bulleted';

const shareComponent = props => (
  <TabList {...props}>
    <Tab>Tab1</Tab>
    <Tab>Tab2</Tab>
  </TabList>
);

test('render TabList', () => {
  const component = mount(
    <TabList>
      <Tab>Tab1</Tab>
    </TabList>
  );
  expect(toJson(component)).toMatchSnapshot();
})

test('custom ListStyle style', () => {
  const CustomListStyle = ListStyle.extend`
    background-color: red;
  `;
  const component = mount(
    <TabList customStyle={CustomListStyle}>
      <Tab>Tab1</Tab>
    </TabList>
  );
  expect(toJson(component)).toMatchSnapshot();  
})

describe('showModalButton', () => {
  const returnMountedDisplay = showModalButton => {
    const component = mount(shareComponent({showModalButton}));
    const buttonWrapper = component.find(MdFormatListBulleted).closest('div');
    return buttonWrapper.instance().style.display;
  }

  it('default: true => show', () => {
    expect(returnMountedDisplay(true)).toEqual('block');
  })
  it('false', () => {
    expect(returnMountedDisplay(false)).toEqual('none');
  })
  describe('number', () => {
    it('2', () => {
      expect(returnMountedDisplay(2)).toEqual('block');
    })
    it('100', () => {
      expect(returnMountedDisplay(100)).toEqual('none');
    })
  })
})