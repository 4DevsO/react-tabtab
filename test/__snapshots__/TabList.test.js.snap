// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom ListStyle style 1`] = `
.c0 {
  position: relative;
  white-space: nowrap;
  overflow: hidden;
  width: auto;
  border-bottom: 1px solid #eee;
  padding: 0;
}

.c1 {
  display: inline-block;
  -webkit-filter: none;
  filter: none;
  position: absolute;
  left: 0;
}

.c1:hover {
  cursor: pointer;
}

.c2 {
  display: inline-block;
  -webkit-filter: none;
  filter: none;
  position: absolute;
  left: 10px;
}

.c2:hover {
  cursor: pointer;
}

.c3 {
  display: inline-block;
  -webkit-filter: none;
  filter: none;
  position: absolute;
  right: 0;
}

.c3:hover {
  cursor: pointer;
}

.c5 {
  padding-left: 0;
  position: relative;
  margin: 0;
  list-style: none;
  display: inline-block;
  -webkit-transition: -webkit-transform .3s cubic-bezier(.42,0,.58,1);
  -webkit-transition: transform .3s cubic-bezier(.42,0,.58,1);
  transition: transform .3s cubic-bezier(.42,0,.58,1);
}

.c6 {
  display: inline-block;
  color: black;
  border-bottom: 0;
  padding: 5px 10px;
  -webkit-transition: color .3s cubic-bezier(.645,.045,.355,1);
  transition: color .3s cubic-bezier(.645,.045,.355,1);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c6:hover {
  cursor: pointer;
  color: #007bff;
}

.c4 {
  overflow: hidden;
  background-color: red;
}

<TabList
  customStyle={[Function]}
>
  <div>
    <styled.div
      hasExtraButton={false}
    >
      <div
        className="c0"
      >
        <styled.div
          innerRef={[Function]}
          onClick={[Function]}
        >
          <div
            className="c1"
            onClick={[Function]}
          >
            <MdFormatListBulleted>
              <IconBase
                viewBox="0 0 40 40"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style={
                    Object {
                      "color": undefined,
                      "verticalAlign": "middle",
                    }
                  }
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m11.6 8.4h23.4v3.2h-23.4v-3.2z m0 13.2v-3.2h23.4v3.2h-23.4z m0 10v-3.2h23.4v3.2h-23.4z m-5-3.8q1 0 1.6 0.6t0.7 1.6-0.7 1.6-1.6 0.6-1.5-0.6-0.6-1.6 0.6-1.6 1.5-0.6z m0-20.3q1.1 0 1.8 0.7t0.7 1.8-0.7 1.8-1.8 0.7-1.7-0.7-0.8-1.8 0.8-1.8 1.7-0.7z m0 10q1.1 0 1.8 0.7t0.7 1.8-0.7 1.8-1.8 0.7-1.7-0.7-0.8-1.8 0.8-1.8 1.7-0.7z"
                    />
                  </g>
                </svg>
              </IconBase>
            </MdFormatListBulleted>
          </div>
        </styled.div>
        <div
          style={
            Object {
              "display": "none",
            }
          }
        >
          <styled.div
            left={true}
            onClick={[Function]}
          >
            <div
              className="c2"
              onClick={[Function]}
            >
              <MdChevronLeft>
                <IconBase
                  viewBox="0 0 40 40"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    preserveAspectRatio="xMidYMid meet"
                    style={
                      Object {
                        "color": undefined,
                        "verticalAlign": "middle",
                      }
                    }
                    viewBox="0 0 40 40"
                    width="1em"
                  >
                    <g>
                      <path
                        d="m25.7 12.3l-7.7 7.7 7.7 7.7-2.3 2.3-10-10 10-10z"
                      />
                    </g>
                  </svg>
                </IconBase>
              </MdChevronLeft>
            </div>
          </styled.div>
          <styled.div
            onClick={[Function]}
          >
            <div
              className="c3"
              onClick={[Function]}
            >
              <MdChevronRight>
                <IconBase
                  viewBox="0 0 40 40"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    preserveAspectRatio="xMidYMid meet"
                    style={
                      Object {
                        "color": undefined,
                        "verticalAlign": "middle",
                      }
                    }
                    viewBox="0 0 40 40"
                    width="1em"
                  >
                    <g>
                      <path
                        d="m16.6 10l10 10-10 10-2.3-2.3 7.7-7.7-7.7-7.7z"
                      />
                    </g>
                  </svg>
                </IconBase>
              </MdChevronRight>
            </div>
          </styled.div>
        </div>
        <styled.div
          innerRef={[Function]}
        >
          <div
            className="c4"
          >
            <styled.div
              innerRef={[Function]}
            >
              <div
                className="c5"
              >
                <Tab
                  active={false}
                  index={0}
                  key="0/.0"
                  tabIndex={0}
                >
                  <styled.div
                    active={false}
                    innerRef={[Function]}
                    onClick={[Function]}
                  >
                    <div
                      className="c6"
                      onClick={[Function]}
                    >
                      Tab1
                    </div>
                  </styled.div>
                </Tab>
              </div>
            </styled.div>
          </div>
        </styled.div>
      </div>
    </styled.div>
  </div>
</TabList>
`;

exports[`render TabList 1`] = `
.c0 {
  position: relative;
  white-space: nowrap;
  overflow: hidden;
  width: auto;
  border-bottom: 1px solid #eee;
  padding: 0;
}

.c1 {
  display: inline-block;
  -webkit-filter: none;
  filter: none;
  position: absolute;
  left: 0;
}

.c1:hover {
  cursor: pointer;
}

.c2 {
  display: inline-block;
  -webkit-filter: none;
  filter: none;
  position: absolute;
  left: 10px;
}

.c2:hover {
  cursor: pointer;
}

.c3 {
  display: inline-block;
  -webkit-filter: none;
  filter: none;
  position: absolute;
  right: 0;
}

.c3:hover {
  cursor: pointer;
}

.c4 {
  overflow: hidden;
}

.c5 {
  padding-left: 0;
  position: relative;
  margin: 0;
  list-style: none;
  display: inline-block;
  -webkit-transition: -webkit-transform .3s cubic-bezier(.42,0,.58,1);
  -webkit-transition: transform .3s cubic-bezier(.42,0,.58,1);
  transition: transform .3s cubic-bezier(.42,0,.58,1);
}

.c6 {
  display: inline-block;
  color: black;
  border-bottom: 0;
  padding: 5px 10px;
  -webkit-transition: color .3s cubic-bezier(.645,.045,.355,1);
  transition: color .3s cubic-bezier(.645,.045,.355,1);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c6:hover {
  cursor: pointer;
  color: #007bff;
}

<TabList>
  <div>
    <styled.div
      hasExtraButton={false}
    >
      <div
        className="c0"
      >
        <styled.div
          innerRef={[Function]}
          onClick={[Function]}
        >
          <div
            className="c1"
            onClick={[Function]}
          >
            <MdFormatListBulleted>
              <IconBase
                viewBox="0 0 40 40"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style={
                    Object {
                      "color": undefined,
                      "verticalAlign": "middle",
                    }
                  }
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m11.6 8.4h23.4v3.2h-23.4v-3.2z m0 13.2v-3.2h23.4v3.2h-23.4z m0 10v-3.2h23.4v3.2h-23.4z m-5-3.8q1 0 1.6 0.6t0.7 1.6-0.7 1.6-1.6 0.6-1.5-0.6-0.6-1.6 0.6-1.6 1.5-0.6z m0-20.3q1.1 0 1.8 0.7t0.7 1.8-0.7 1.8-1.8 0.7-1.7-0.7-0.8-1.8 0.8-1.8 1.7-0.7z m0 10q1.1 0 1.8 0.7t0.7 1.8-0.7 1.8-1.8 0.7-1.7-0.7-0.8-1.8 0.8-1.8 1.7-0.7z"
                    />
                  </g>
                </svg>
              </IconBase>
            </MdFormatListBulleted>
          </div>
        </styled.div>
        <div
          style={
            Object {
              "display": "none",
            }
          }
        >
          <styled.div
            left={true}
            onClick={[Function]}
          >
            <div
              className="c2"
              onClick={[Function]}
            >
              <MdChevronLeft>
                <IconBase
                  viewBox="0 0 40 40"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    preserveAspectRatio="xMidYMid meet"
                    style={
                      Object {
                        "color": undefined,
                        "verticalAlign": "middle",
                      }
                    }
                    viewBox="0 0 40 40"
                    width="1em"
                  >
                    <g>
                      <path
                        d="m25.7 12.3l-7.7 7.7 7.7 7.7-2.3 2.3-10-10 10-10z"
                      />
                    </g>
                  </svg>
                </IconBase>
              </MdChevronLeft>
            </div>
          </styled.div>
          <styled.div
            onClick={[Function]}
          >
            <div
              className="c3"
              onClick={[Function]}
            >
              <MdChevronRight>
                <IconBase
                  viewBox="0 0 40 40"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    preserveAspectRatio="xMidYMid meet"
                    style={
                      Object {
                        "color": undefined,
                        "verticalAlign": "middle",
                      }
                    }
                    viewBox="0 0 40 40"
                    width="1em"
                  >
                    <g>
                      <path
                        d="m16.6 10l10 10-10 10-2.3-2.3 7.7-7.7-7.7-7.7z"
                      />
                    </g>
                  </svg>
                </IconBase>
              </MdChevronRight>
            </div>
          </styled.div>
        </div>
        <styled.div
          innerRef={[Function]}
        >
          <div
            className="c4"
          >
            <styled.div
              innerRef={[Function]}
            >
              <div
                className="c5"
              >
                <Tab
                  active={false}
                  index={0}
                  key="0/.0"
                  tabIndex={0}
                >
                  <styled.div
                    active={false}
                    innerRef={[Function]}
                    onClick={[Function]}
                  >
                    <div
                      className="c6"
                      onClick={[Function]}
                    >
                      Tab1
                    </div>
                  </styled.div>
                </Tab>
              </div>
            </styled.div>
          </div>
        </styled.div>
      </div>
    </styled.div>
  </div>
</TabList>
`;
